In [4]:
# Structural Analysis of the disulfide bonds in 2q7q
# Author: Eric G. Suchanek, PhD.
# (c) 2023 Eric G. Suchanek, PhD., All Rights Reserved
# Last Modification: 2/2/24
# Cα N, Cα, Cβ, C', Sγ Å °

import pandas as pd
import pyvista as pv
from pyvista import set_plot_theme

from proteusPy.Disulfide import *
from proteusPy.DisulfideList import DisulfideList
from proteusPy.DisulfideLoader import Load_PDB_SS

# pyvista setup for notebooks
pv.set_jupyter_backend('trame')

set_plot_theme('dark')
LIGHT = True

# The Most Beautiful Disulfide Bond in the World
Eric G. Suchanek, PhD 2/14/23



### Load the RCSB Disulfide Database
We load the database and get its properties as follows:

In [2]:
PDB_SS = Load_PDB_SS(verbose=True)
PDB_SS.describe()

-> load_PDB_SS(): Reading /Users/egs/miniforge3/envs/proteusPy/lib/python3.11/site-packages/proteusPy/data/PDB_SS_ALL_LOADER.pkl... 
-> load_PDB_SS(): Done reading /Users/egs/miniforge3/envs/proteusPy/lib/python3.11/site-packages/proteusPy/data/PDB_SS_ALL_LOADER.pkl... 
    =========== RCSB Disulfide Database Summary ==============
       =========== Built: 2024-02-08 16:06:42 ==============
PDB IDs present:                    35818
Disulfides loaded:                  120494
Average structure resolution:       2.34 Å
Lowest Energy Disulfide:            2q7q_75D_140D
Highest Energy Disulfide:           1toz_456A_467A
Cα distance cutoff:                 8.00 Å
Total RAM Used:                     30.72 GB.
    ================= proteusPy: 0.89 =======================


We see from the statistics above that disulfide 2q7q_75D_140D has the lowest energy, so let's extract it from the database and display it. 

A few notes about the display window. You might need to click into the window to refresh it. Click drag to rotate the structures, mousewheel to zoom. The window titles display several parameters about the disulfide bonds including their approximate torsional energy, their Ca-Ca distance, and the *torsion length*. 

The latter parameter is formally, the Euclidean length of the sidechain dihedral angle when treated as a five-dimensional vector. This sounds all mathy and complicated, but in essence it gives a measure of how 'long' that five dimensional vector is. This is used by the DisulfideBond routines to compare individual structures and gauge their structural similarity.

In [5]:
best_ss = PDB_SS['2q7q_75D_140D']
best_ss.pprint()
best_ss.display(style='sb', light=LIGHT)

<Disulfide 2q7q_75D_140D, Source: 2q7q, Resolution: 1.6 Å 
Χ1-Χ5: -59.36°, -59.28°, -83.66°, -59.82° -59.91°, -25.17°, 0.49 kcal/mol 
Cα Distance: 5.50 Å 
Torsion length: 145.62 deg>


Widget(value='<iframe src="http://localhost:58240/index.html?ui=P_0x280070510_1&reconnect=auto" class="pyvista…

And that, gentle reader, is it. *The most beautiful disulfide bond in the world*! Look at it. It's gorgeous. The sidechain dihdedral angles (Χ1-Χ5: -59.36°, -59.28°, -83.66°, -59.82° -59.91°), and estimated energy, (0.49 kcal/mol) is *at* than the analytical global minimum, (Χ1-Χ5: -60.00°, -60.00°, -90.00°, -60.00° -60.00°, 0.60 kcal/mol)! Let's have a look at the 'CPK' style rendering to see the structure's overall appearance:


In [4]:
best_ss.display(style='cpk', light=LIGHT)

Widget(value='<iframe src="http://localhost:53233/index.html?ui=P_0x28a4be150_1&reconnect=auto" class="pyvista…

### Finding Potential Structural Relatives by Sequence Similarity
Working under the assumption that similar sequence -> similar structure I generated a query on: https://www.ebi.ac.uk/pdbe/entry/pdb/2q7q to return PDB IDs for structures with high sequence similarity. This yielded a ```.csv``` file, which we will import below:

In [5]:
ss_df = pd.read_csv('2q7q_seqsim.csv')
ss_df.head(5)

pdb_id  organism_scientific_name  tax_id  \
0   2q7q  Paracoccus denitrificans     266   
1   2bbk  Paracoccus denitrificans     266   
2   2agy      Alcaligenes faecalis     511   
3   2agy      Alcaligenes faecalis     511   
4   2ah1      Alcaligenes faecalis     511   

                                   organism_synonyms  \
0  Parde,Paracoccus Denitrificans,Micrococcus Den...   
1  Parde,Paracoccus Denitrificans,Micrococcus Den...   
2  Achromobacter Sp. Atcc8750,Alcaligenes Sp. Bp1...   
3  Achromobacter Sp. Atcc8750,Alcaligenes Sp. Bp1...   
4  Achromobacter Sp. Atcc8750,Alcaligenes Sp. Bp1...   

                                                rank        genus  \
0  species,genus,family,order,class,phylum,superk...   Paracoccus   
1  species,genus,family,order,class,phylum,superk...   Paracoccus   
2  species,genus,family,order,class,phylum,superk...  Alcaligenes   
3  species,genus,family,order,class,phylum,superk...  Alcaligenes   
4  species,genus,family,order,class,phylum,superk...  Alcaligenes   

  superkingdom        journal  journal_volume  journal_first_page  ...  \
0     Bacteria  J. Mol. Biol.           276.0                 NaN  ...   
1     Bacteria  J. Mol. Biol.           276.0                 NaN  ...   
2     Bacteria        Science           312.0                 NaN  ...   
3     Bacteria        Science           312.0                 NaN  ...   
4     Bacteria        Science           312.0                 NaN  ...   

                            molecule_name all_molecule_name  \
0   Methylamine dehydrogenase heavy chain               NaN   
1   Methylamine dehydrogenase light chain               NaN   
2  Aralkylamine dehydrogenase light chain               NaN   
3  Aralkylamine dehydrogenase heavy chain               NaN   
4  Aralkylamine dehydrogenase light chain               NaN   

  modified_residue_flag molecule_type  mutation_type  entry_uniprot_accession  \
0                     N       Protein       Conflict            P29894,P22619   
1                     Y       Protein            NaN            P29894,P22619   
2                     Y       Protein            NaN            P84887,P84888   
3                     N       Protein            NaN            P84887,P84888   
4                     Y       Protein            NaN            P84888,P84887   

   uniprot_id                                   molecule_synonym gene_name  \
0  DHMH_PARDE  Methylamine dehydrogenase (amicyanin),Methylam...      mauB   
1  DHML_PARDE  Methylamine dehydrogenase (amicyanin),MADH,Met...      mauA   
2  AAUA_ALCFA  Aromatic amine dehydrogenase,AADH,Aralkylamine...      aauA   
3  AAUB_ALCFA  Aromatic amine dehydrogenase,Aralkylamine dehy...      aauB   
4  AAUA_ALCFA  Aromatic amine dehydrogenase,AADH,Aralkylamine...      aauA   

  entity_id  
0         1  
1         2  
2         1  
3         2  
4         1  

[5 rows x 59 columns]

All of the nearest sequence neighbors are sadly, bacterial. Let's extract the unique ids next.

In [6]:
relative_list = ss_df['pdb_id'].unique()
relative_list

array(['2q7q', '2bbk', '2agy', '2ah1', '2ah0', '2agl', '2agx', '2hjb',
       '1mae', '2oiz', '2ojy', '2i0s', '2iup', '2iur', '2agw', '2hxc',
       '2i0r', '2iuv', '2i0t', '2mad', '2agz', '2hkr', '2hj4', '2ok4',
       '2hkm', '1maf', '2ok6', '2iuq', '3orv', '2h47', '2h3x', '3l4m',
       '3l4o', '2j57', '2j55', '2j56', '3pxt', '3sle', '3c75', '3rn0',
       '3sjl', '3pxw', '3pxs', '3rlm', '3rmz', '4fa1', '4fa9', '3sxt',
       '4l3h', '3rn1', '3sws', '4o1q', '4l3g', '4k3i', '4l1q', '4fan',
       '4fa5', '4fav', '4fb1', '3svw', '4y5r', '4fa4', '2iaa', '1mg3',
       '1mg2', '2gc4', '2gc7', '2mta'], dtype=object)

We now need to convert the list of PDB IDs into real disulfides from the database. We do this with the ``DisulfideLoader.build_ss_from_idlist()`` function. Next we print out some relevant statistics.


In [7]:
relatives = DisulfideList([], 'relatives')
relatives = PDB_SS.build_ss_from_idlist(relative_list)

print(f'There are: {relatives.length} related structures.\nAverage Energy: {relatives.Average_Energy:.2f} kcal/mol\nAverage Ca distance: {relatives.Average_Distance:.2f} Å')
print(f'Average resolution: {relatives.Average_Resolution:.2f} Å \nAverage torsion distance: {relatives.Average_Torsion_Distance:.2f}°')

There are: 68 related structures.
Average Energy: 2.23 kcal/mol
Average Ca distance: 3.17 Å
Average resolution: 1.88 Å 
Average torsion distance: 123.83°


In [ ]:
ssmin, ssmax = relatives.minmax_energy()
duolist = DisulfideList([ssmin, ssmax], 'mM')
duolist.display(style='sb', light=LIGHT)

In [ ]:
duolist.display_overlay(light=LIGHT)

The two Disulfides...

We can find disulfides that are conformationally related by using the DisulfideList.nearest_neighbors() function with a dihedral angle cutoff. This cutoff is measure of angular similarity across all five sidechain dihedral angles.

In [ ]:
close_neighbors = relatives.nearest_neighbors(ssmin.chi1, ssmin.chi2, ssmin.chi3, ssmin.chi4, ssmin.chi5, 10.0)
close_neighbors.length

In [ ]:
close_neighbors.display_overlay(light=LIGHT)

In [ ]:
ssTotList = PDB_SS.SSList
global_neighbors = ssTotList.nearest_neighbors(ssmin.chi1, ssmin.chi2, ssmin.chi3, ssmin.chi4, ssmin.chi5, 5.0)
global_neighbors.length

In [ ]:
global_neighbors.display_overlay(light=LIGHT)

In [ ]:
from scipy.optimize import minimize
import numpy as np


def ss_energy_function(x):
    chi1, chi2, chi3, chi4, chi5 = x
    energy = 2.0 * (np.cos(np.deg2rad(3.0 * chi1)) + np.cos(np.deg2rad(3.0 * chi5)))
    energy += np.cos(np.deg2rad(3.0 * chi2)) + np.cos(np.deg2rad(3.0 * chi4))
    energy += 3.5 * np.cos(np.deg2rad(2.0 * chi3)) + 0.6 * np.cos(np.deg2rad(3.0 * chi3)) + 10.1
    return energy
    
initial_guess = [-60.0, -60.0, -90.0, -60.0, -60.0] # initial guess for chi1, chi2, chi3, chi4, chi5
result = minimize(ss_energy_function, initial_guess, method="Nelder-Mead")
minimum_energy = result.fun
inputs = result.x

print(f'Minimum Energy: {minimum_energy:.3f} for conformation: {[f"{x:.3f}" for x in inputs]}')


## References
* *Application of Artificial Intelligence in Protein Design* - Doctoral Dissertation, EG Suchanek, 1987, Johns Hopkins Medical School
* https://doi.org/10.1021/bi00368a023
* https://doi.org/10.1021/bi00368a024
* https://doi.org/10.1016/0092-8674(92)90140-8
* http://dx.doi.org/10.2174/092986708783330566
* https://doi.org/10.1021/bi0603064
* https://doi.org/10.1021/bi9826658
* https://pubmed.ncbi.nlm.nih.gov/22782563/
* 